test/LocalVocabTest.cpp

```diff
@@ -20,23 +34,146 @@ TEST(LocalVocab, construction) {
   // Create empty local vocabulary.
   LocalVocab localVocab;
   ASSERT_TRUE(localVocab.empty());
+
   // Add the words from our test vocabulary and check that they get the expected
   // local IDs.
   for (size_t i = 0; i < testVocab.size(); ++i) {
     ASSERT_EQ(localVocab.getIndexAndAddIfNotContained(testVocab[i]),
               LocalVocabIndex::make(i));
   }
   ASSERT_EQ(localVocab.size(), testVocab.size());
+
   // Check that we get the same IDs if we do this again, but that no new words
   // will be added.
   for (size_t i = 0; i < testVocab.size(); ++i) {
     ASSERT_EQ(localVocab.getIndexAndAddIfNotContained(testVocab[i]),
               LocalVocabIndex::make(i));
   }
   ASSERT_EQ(localVocab.size(), testVocab.size());
+
   // Check that the lookup by ID gives the correct words.
   for (size_t i = 0; i < testVocab.size(); ++i) {
     ASSERT_EQ(localVocab.getWord(LocalVocabIndex::make(i)), testVocab[i]);
   }
   ASSERT_EQ(localVocab.size(), testVocab.size());
+
+  // Check that out-of-bound lookups are detected (this would have a caught the
+  // one-off bug in #820, LocalVocab.cpp line 55).
+  ASSERT_THROW(localVocab.getWord(LocalVocabIndex::make(testVocab.size())),
+               std::runtime_error);
+  ASSERT_THROW(localVocab.getWord(LocalVocabIndex::make(-1)),
+               std::runtime_error);
+
+  // Check that a move gives the expected result.
+  auto localVocabMoved = std::move(localVocab);
+  ASSERT_EQ(localVocab.size(), 0);
+  ASSERT_EQ(localVocabMoved.size(), testVocab.size());
+  for (size_t i = 0; i < testVocab.size(); ++i) {
+    ASSERT_EQ(localVocabMoved.getWord(LocalVocabIndex::make(i)), testVocab[i]);
+  }
+}
+
+// _____________________________________________________________________________
+TEST(LocalVocab, propagation) {
+  // Query execution context (with small test index) and allocator for testing,
+  // see `IndexTestHelpers.h`.
+  using ad_utility::AllocatorWithLimit;
+  QueryExecutionContext* testQec = ad_utility::testing::getQec();
+  AllocatorWithLimit<Id> testAllocator = ad_utility::testing::makeAllocator();
+
+  // Lambda that checks the contents of the local vocabulary after the specified
+  // operation.
+  // NOTE: Clear the cache before each check
+  // TODO: Does this solve the problem with the local vocab when doing several
+  // operations on `values1`?
+  auto checkLocalVocab = [&](auto* operation,
+                             std::vector<std::string> expectedWords) -> void {
+    operation->getExecutionContext()->getQueryTreeCache().clearAll();
+    ResultTable resultTable(testAllocator);
+    operation->computeResult(&resultTable);
```

Comment:
You asked about the `computeResult` method being `private`.
If you always clear the cache, then you can use `getResult()`, which is public.